### PR TITLE
Add strict comparison sniff

### PR DESCRIPTION
**Pull request by VincentLanglet**
WIP for https://github.com/squizlabs/PHP_CodeSniffer/issues/2238

### CODE REVIEW

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @return array
+     */
+    public function register()
+    {
+        return [
+            T_IS_EQUAL,
```
Comment (jrfnl) [inline src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: What's the point of having the public property if you don't use it in `register()` ? People could now add additional operators via the public property, but they would never be sniffed for.

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+     *                                               stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+
+        $phpcsFile->addError(
+            'The %s comparator is forbidden, use %s instead',
```
Comment (jrfnl) [inline src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: `comparator` is an electronic circuit.

```suggestion
            'Using the %s operator is forbidden, use %s instead',
```

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Ensures that constant names are all uppercase.
+ *
+ * @author    Vincent Langlet <vincentlanglet@exemple.com>
+ * @copyright 2006-2015 Squiz Pty Ltd (ABN 77 084 670 600)
```
Comment (jrfnl) [inline src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]:
```suggestion
 * @copyright 2020 Squiz Pty Ltd (ABN 77 084 670 600)
```

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * Ensures that constant names are all uppercase.
```
Comment (jrfnl) [inline src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: Copy/pasta.

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+
+        $phpcsFile->addError(
+            'The %s comparator is forbidden, use %s instead',
+            $stackPtr,
+            'NotStrict',
```
Comment (jrfnl) [inline src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]:
```suggestion
            'Forbidden',
```

File: src/Standards/Generic/Tests/Operators/StrictComparisonUnitTest.inc [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+if ($true == true) {
+    echo 'True';
+} elseif ($true != true) {
+    echo 'False';
+} elseif (false === $true && true !== $false) {
+    echo 'False';
+}
+
+$a = $b == $c;
+$a = $b != $c;
```
Comment (jrfnl) [inline src/Standards/Generic/Tests/Operators/StrictComparisonUnitTest.inc]: Please add tests which test the public property as well.

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+class StrictComparisonSniff implements Sniff
+{
+
+    /**
+     * Types to replace: key is operator to replace, value is operator to replace with.
+     *
+     * @var array
+     */
+    public $operators = [
+        T_IS_EQUAL     => '===',
```
Comment (jrfnl) [inline src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: Using the constant as the key will make this property nearly impossible to use as the value of the token constants may change between PHP versions.

Suggestion: use the type instead, i.e `'T_IS_EQUAL'` (as a string) to allow for setting this from a ruleset in a reliable manner.

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: Whoah, I'm tired. Thanks.

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: How can I get the code in the register method from the type then ?

Comment (jrfnl) [reply src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: You mean this ?

```php
$targets = [];
foreach ($this->operators as $type => $replacement) {
    $targets[] = constant($type);
}

return $targets;
```

**Comment by VincentLanglet** [comment]
@jrfnl I'm not sure about the "generic" `ForbiddenOperators` (or even `ForbiddenTokens`). It does lose context and it forces the user to configure this Sniff. There is already a `ValidLogicalOperatorsSniff` and the `BacktickOperatorSniff` for example and I think there're better than a Generic. But an AbstractForbiddenTokensSniff can be made, and they can extends it.

After some thinking, I feel like the $forbiddenOperator option should not be even needed.
The issue with the `ComparisonOperatorUsageSniff` was the fact this sniff check multiple things in one sniff and some people wanted to remove one or multiple checks (never adding one).
By splitting this Sniff in 3 minimal sniff, 
- One for === and !==.
- One for ! usage.
- One for implicit True.
you can take exactly what you want.

File: src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Standards/Generic/Sniffs/Operators/StrictComparisonSniff.php]: Yes, thanks, I didn't know the constant function.

**Pull request closed without merging**